diff --git a/iongraph b/iongraph
index 59fc9e6..b39a367 100755
--- a/iongraph
+++ b/iongraph
@@ -132,17 +132,23 @@ def getInstructionRow(inst):
         instType = '<td align="left">%s</td>' % cgi.escape(inst['type'])
 
     return '<tr>%s%s%s</tr>' % (instLabel, instName, instType)
 
 # block obj -> HTML-formatted string
 def getBlockLabel(b):
     s =  '<<table border="0" cellborder="0" cellpadding="1">'
 
-    blockTitle = '<font color="white">Block %s</font>' % str(b['number'])
+    if 'blockUseCount' in b:
+        blockUseCount = str(b['blockUseCount'])
+    else:
+        blockUseCount = 'N/A'
+
+    blockTitle = '<font color="white">Block %s (Count: %s)</font>' % \
+            (str(b['number']), blockUseCount)
     blockTitle = '<td align="center" bgcolor="black" colspan="3">%s</td>' % blockTitle
     s += '<tr>%s</tr>' % blockTitle
     
     if 'resumePoint' in b:
         s += getResumePointRow(b['resumePoint'], None)
 
     for inst in b['instructions']:
         if 'resumePoint' in inst:
